client/blocks/support-article-dialog/dialog.jsx

```diff
@@ -1,55 +1,60 @@
 import getCurrentQueryArguments from 'calypso/state/selectors/get-current-query-arguments';
 import getInlineSupportArticleActionIsExternal from 'calypso/state/selectors/get-inline-support-article-action-is-external';
 import getInlineSupportArticleActionLabel from 'calypso/state/selectors/get-inline-support-article-action-label';
 import getInlineSupportArticleActionUrl from 'calypso/state/selectors/get-inline-support-article-action-url';
 import getInlineSupportArticleBlogId from 'calypso/state/selectors/get-inline-support-article-blog-id';
 import getInlineSupportArticlePostId from 'calypso/state/selectors/get-inline-support-article-post-id';
-import {
-	getSupportArticleAlternatesForLocale,
-	isRequestingSupportArticleAlternates,
-	shouldRequestSupportArticleAlternates,
-} from 'calypso/state/support-articles-alternates/selectors';
 import SupportArticleHeader from './header';
 import Placeholders from './placeholders';
 
 import './style.scss';
 import './content.scss';
 
 const noop = () => {};
+const getPostKey = memoize(
+	( blogId, postId ) => ( { blogId, postId } ),
+	( ...args ) => JSON.stringify( args )
+);
 
-export const SupportArticleDialog = ( {
-	actionIsExternal,
-	actionLabel,
-	actionUrl,
-	closeSupportArticleDialog,
-	post,
-	postId,
-	postKey,
-	shouldRequestAlternates,
-	isRequestingAlternates,
-} ) => {
+export const SupportArticleDialog = () => {
 	const translate = useTranslate();
-	const isLoading = ! post || isRequestingAlternates;
-	const siteId = post?.site_ID;
-	const shouldQueryReaderPost = ! post && ! shouldRequestAlternates && ! isRequestingAlternates;
 	const { value: supportArticleId, closeModal } = useRouteModal( 'support-article' );
+	const actionUrl = useSelector( getInlineSupportArticleActionUrl );
+	const actionLabel = useSelector( getInlineSupportArticleActionLabel );
+	const actionIsExternal = useSelector( getInlineSupportArticleActionIsExternal );
 	const articleUrl = actionUrl ? actionUrl : 'https://support.wordpress.com?p=' + supportArticleId;
+	const requestBlogId = useSelector( getInlineSupportArticleBlogId );
+	const blogId = requestBlogId ?? SUPPORT_BLOG_ID;
+	const currentQueryArgs = useSelector( getCurrentQueryArguments );
+	let postId = useSelector( getInlineSupportArticlePostId );
+	if ( ! postId ) {
+		postId = parseInt( currentQueryArgs?.[ 'support-article' ], 10 );
+	}
+	let postKey = getPostKey( blogId, postId );
+	const supportArticleAlternates = useSupportArticleAlternatesQuery( blogId, postId );
+	if ( supportArticleAlternates.data ) {
+		postKey = getPostKey(
+			supportArticleAlternates.data.blog_id,
+			supportArticleAlternates.data.page_id
+		);
+	}
+	const post = useSelector( ( state ) => getPostByKey( state, postKey ) );
+	const isLoading = ! post || supportArticleAlternates.isFetching;
+	const siteId = post?.site_ID;
+	const shouldQueryReaderPost = ! post && ! supportArticleAlternates.isFetching;
```

Comment:
The alternates logic could be extracted to a hook:
```js
const postKey = useSupportArticleAlternatePostKey( blogId, postId );
```
It would return `null` when alternates are loading, key for the original `blogId+postId` if the query doesn't find an alternate, and alternate key if it was found and loaded.

In the current logic you have situations where there's a valid key in the `postKey` local variable, but you must be careful not to use it, because the alternate is not loaded yet.

Reply:
Good eye and a great suggestion. Moved post key logic to a hook in 9cbc5b85f29d09d058742ae96ce78ac40312a049.